### PR TITLE
Security release process description

### PR DESCRIPTION
**Pull request by fturib**
### 1. What does this pull request do?

Add a sub-section in the section "Security" of README, describing the security release process that CoreDNS expect to follow  in order to be compatible with K8s's own security release process

### 2. Which issues (if any) are related?
CoreDNS going GA for k8s.

### 3. Which documentation changes (if any) need to be made?
README.

### CODE REVIEW

**Comment by fturib** [comment]
@miekg , @johnbelamaric , following recommendation of Tim A. from K8s Team, I propose that we have a description of our security release process.

Right now, I just enhance the README.

Let me know if that is preferable to have a separated document ... and where.

Initial reco from Tim A .is:

> I think the requirements for the CoreDNS security process to be compatible with ours are:
> 
> 1. Disclosures can be made privately
> 2. kubernetes-security@googlegroups.com gets early notification of upcoming security releases
> 3. A compatible public disclosure & release timeline, so we have enough time to coordinate a corresponding Kubernetes release. Specifically:
>   - public releases should correspond with public disclosure (controlled), i.e. within O(hours)
>   - our timelines need to be revisited, but ideally we would have ~1 week to pull the release together. We could probably scramble in 24 hours for a critical emergency.
> 4. Backwards-compatible patch releases are provided for releases going back at least as far as the Kubernetes support timeline (3 releases).

Let me know what you think.

Thanks

File: README.md
Comment (miekg) [inline README.md]: I have some opinions about this

File: README.md
Comment (miekg) [inline README.md]: Which is nice and all, but does that more it more special than, say, Cloudflare which also use coredns in some project?

File: README.md
Comment (miekg) [inline README.md]: The paragraph above also talks about the security process, should probably merge these two

File: README.md
Comment (miekg) [inline README.md]: We should make this generic and create a private security announce where interesting parties should subscribe

File: README.md
Comment (miekg) [inline README.md]: We have no such policy and I think the unofficial one is "upgrade to latest version". I also believe we lack the people power to provide such backwards compatibility.

File: README.md
Comment (fturib) [reply README.md]: Good point. I agree.

File: README.md
Comment (fturib) [reply README.md]: As you propose below, let's make something more generic with "collaborating projects"

File: README.md
Comment (fturib) [reply README.md]: So, if we have to provide a security release, it will be on top of the current official release ? 
for instance, right now, it would be on top of 1.1.3.

The collaborating projects would have to release a patch that includes that new release "1.1.3-security-patch-01" ?

**Comment by fturib** [comment]
I pushed a second version of the "Security releases coordination"

File: README.md
Comment (fturib) [inline README.md]: Who has a better name for "Collaborating Projects" ? (maybe included Security in that name ..) ?

File: README.md
Comment (fturib) [inline README.md]: In order to be transparent and for each project to be able to validate they are registered, should we create a public document CollaboratingProjects.md  with the public information regarding process and agreement with each project ?

File: README.md
Comment (fturib) [inline README.md]: That last line is vague in purpose and let room for agreement on what versions will be patched - based on the needs defined during "coordination" of the security release.
NOTE: that is far from K8s expectation right now.

File: README.md
Comment (miekg) [reply README.md]: I think we should do the opposite and require any interested project to join (a new) mailing list

Comment (fturib) [reply README.md]: you mean the list of "Collaborating Projects" should be a simple mailing list ?
I guess we need to verify that the subscribing project is really using CoreDNS.. and mostly will keep private that information until the release, no ?

So, is that ok that any new project subscribe to be on the mailing list by sending request through 'secure@coredns.io' ?

then we validate and add this project to the mailing list - let say dl-security-release@coredns.io,

Comment (fturib) [reply README.md]: I guess, in practical, what would happen is:
1- we receive a vulnerability issue somehow (incl email at security@coredns.io)
2- we assess if that needs a security release of CoreDNS
if YES:
3- we (CoreDNS Team) elaborate a timeline for release (I guess in a way that is compatible with subscribed projects)
4- we then inform the dl-security-release of the issue and our plan to remediate
5- I guess if that needs bargaining for the release timeframe, then we will be contacted by those projects that expect adaptation of the plan.
6- When plan is definitive, resend an email to advertise that plan and ETA release for CoreDNS.

Would that work ?

**Comment by fturib** [comment]
NOTE: there is a specific document in [k8s about their security release process.](https://github.com/kubernetes/sig-release/blob/master/security-release-process-documentation/security-release-process.md). In that document there is a [specific paragraph on "private distributor list"](https://github.com/kubernetes/sig-release/blob/master/security-release-process-documentation/security-release-process.md#private-distributors-list) in order to coordinate the release.

Should we take that document as template, simplify it, and tune it for CoreDNS project ?

I also found this document on line, edited by CERT :  "[Guide to Coordinated Vulnerability Disclosure](https://resources.sei.cmu.edu/asset_files/SpecialReport/2017_003_001_503340.pdf)"

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP - enhance..." ]
>
>NOTE: there is a specific document in [k8s about their security release process.](https://github.com/kubernetes/sig-release/blob/master/security-release-process-documentation/security-release-process.md). In that document there is a [specific paragraph on "private distributor list"](https://github.com/kubernetes/sig-release/blob/master/security-release-process-documentation/security-release-process.md#private-distributors-list) in order to coordinate the release.
>
>Should we take that document as template, simplify it, and tune it for CoreDNS project ?

I think we should do something like that; yes, probably less heavyweight. One 
issue is that we need to describe when someone is allowed on that list; How do 
we vet such requests?

/Miek

--
Miek Gieben

**Comment by fturib** [comment]
I am working on it.
There is already a description on what is criteria to accept new distributor in the list - need some adaptation though.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP - enhance..." ]
>I am working on it.
>There is already a description on what is criteria to accept new distributor in the list - need some adaptation though.

Ack, cool.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [inline SECURITY-RELEASE-PROCESS.md]: ok.. I expected a table of content here. Is there a way to automate it ?

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [inline SECURITY-RELEASE-PROCESS.md]: John is in PTO .. I could not ask him.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [inline SECURITY-RELEASE-PROCESS.md:100]: Here I tried to describe the process that was used for last security patch.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [inline SECURITY-RELEASE-PROCESS.md]: in k8s, this list is named "Private Distributors List" - not sure what fit the best for CoreDNS ...

**Comment by fturib** [comment]
I uploaded a new version based on existing security release process from k8s project.
I simplified as much as I can. Keeping all the part related to "private integrator list" and cooperation with projects using CoreDNS as a component,

Please feel free to amend, propose change, etc ..

**Comment by fturib** [comment]
I had advises from our Security Department on how to improve this document. 
Will update soon, integrating advices.

**Comment by fturib** [comment]
I update this document:
- includes now a definition of "Product Security Team" (need to define who is part of this Team in Owners.md) 
- a "Fix Team" build-up when a security release has to be made, with a "Team Lead" that have the responsibility to communicate and conduct the release.

NOTE: this new version is a merge of my former version (based on K8s document) and the same process defined in envoy project.

**Comment by yongtang** [comment]
LGTM

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: s/if/If/

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: I think you mean "compressed packet captures"?

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: s/an distributor/a distributor/

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
A few nits, but overall I think the process looks good.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: ok.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: yes.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: ok

**Comment by caniszczyk** [comment]
LGTM, let me know when you want CNCF to create the mailing list

**Comment by fturib** [comment]
@johnbelamaric , @yongtang , @chrisohaver , @miekg : I added you in PST.
Please review and give your agreement (or not).

@miekg : who should create the needed mailing list on cncf.io (see TODO in the DESCRIPTION of this PR, and comment here : https://github.com/coredns/coredns/pull/1861#issuecomment-424373288)

**Comment by fturib** [comment]
@miekg, @johnbelamaric : can you verify you are ok with the latest version ? Thanks.

File: SECURITY-RELEASE-PROCESS.md
Comment (chrisohaver) [inline SECURITY-RELEASE-PROCESS.md]: Language tweak:

"The CoreDNS project has adopted this security disclosure and response policy to ensure responsible handling of critical issues."

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: The CoreDNS project ... to ensure responsible handling of ...

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: s/of set/of the set/

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: Rephrase this to:

If you find a security vulnerability or any security related issues,
please DO NOT file a public issue. Do not create a GitHub issue.
Instead, send your report privately to ...

Also, is that email address supposed to be the same as "product-security-team" or is it really a different address?

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: s/If you know it, please/Please/

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: /forget :/forget:/

File: SECURITY-RELEASE-PROCESS.md
Comment (johnbelamaric) [inline SECURITY-RELEASE-PROCESS.md]: s/actively/active/

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
I bunch of little tweaks in the language, and a question on mailing lists. Maybe there should be a little section that highlights the different mailing lists and their purposes.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: ok

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: done

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: you are right. It can be the same.
security@coredns.io is already existing. So best would be to keep this one.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: done

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: i changed to that existing security@coredns.io email

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: done.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: done

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: done

File: OWNERS
Comment (miekg) [inline OWNERS]: this change can't be made to the OWNERS file just yet, because dreck doesn't support it

File: README.md
Comment (miekg) [inline README.md]: page?

File: SECURITY-RELEASE-PROCESS.md
Comment (miekg) [inline SECURITY-RELEASE-PROCESS.md]: the @coredns.io is a bit adhoc at the moment and requires me manually adding people (or paying $$ for a actual group), this should probably be a closed group on groups.google.com or something hosted by the cnfc we people can subscribe.

File: SECURITY-RELEASE-PROCESS.md
Comment (miekg) [inline SECURITY-RELEASE-PROCESS.md]: Where is the hall of fame?

File: SECURITY-RELEASE-PROCESS.md
Comment (miekg) [reply SECURITY-RELEASE-PROCESS.md]: possibly aliased from distributors-announce@coredns.io

File: SECURITY-RELEASE-PROCESS.md
Comment (miekg) [inline SECURITY-RELEASE-PROCESS.md]: extra whitespace

File: OWNERS
Comment (miekg) [reply OWNERS]: I'll add some thing to dreck.

On second thought; does this need to be public?

**Review by miekg — COMMENTED** [review]
overall lgtm

File: OWNERS
Comment (fturib) [reply OWNERS]: > this change can't be made to the OWNERS file just yet, because dreck doesn't support it
What is dreck ? You mean it needs to be ingested by a tool ?


> On second thought; does this need to be public?
You are right, I do not think it is mandatory to list the persons. But that would be more clear for everyone to know who is in this list and who is not.
NOTE: I just copied that behavior from other project.

File: README.md
Comment (fturib) [reply README.md]: Right ... need to fix that.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Security rele..." ]
>> this change can't be made to the OWNERS file just yet, because dreck doesn't support it
>What is dreck ? You mean it needs to be ingested by a tool ?

github.com/miekg/dreck the thing behind CorBot, yes this file is read by a tool.

>> On second thought; does this need to be public?
>You are right, I do not think it is mandatory to list the persons. But that would be more clear for everyone to know who is in this list and who is not.
>NOTE: I just copied that behavior from other project.

ack.

File: SECURITY-RELEASE-PROCESS.md
Comment (fturib) [reply SECURITY-RELEASE-PROCESS.md]: Do you mean we need to ask creation of the dl for the 2 lists or only for distributor.
Something like:
- coredns-distributor-announce@googlegroup.com (to warn the distributors of early release)
- coredns-security-team@googlegroup.com (to discuss about security patch, only the PST team members)

and keep security@coredns.io for external request on 'security' ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Security rele..." ]
>+* security@coredns.io : for any security concerns. Received by Product Security Team members, and used by this Team to discuss security issues and fixes.
>+* distributors-announce@coredns.io: for early private information on Security patch releases. see below how CoreDNS distributors can apply for this list.
>
>Do you mean we need to ask creation of the dl for the 2 lists or only for distributor.
>Something like:
>- coredns-distributor-announce@googlegroup.com (to warn the distributors of early release)
>- coredns-security-team@googlegroup.com (to discuss about security patch, only the PST team members)
>
>and keep security@coredns.io for external request on 'security' ?

No we can keep security@coredns.io as-is. We do need coredns-distributor-announce@googlegroup.com
and then make a forward from distributors-announce@coredns.io to point to that.
Does that work?

/Miek

--
Miek Gieben

**Comment by fturib** [comment]
- I added a "security" entry in OWNER and corresponding PR for dreck : https://github.com/miekg/dreck/pull/107
- I removed unknown hall of fame
- the mailing list for announcing the early release will be : coredns-distributors-announce@googlegroup.com.

NOTE: when approval is ok. I will ask @caniszczyk if he can to create that private mailing list.

**Comment by fturib** [comment]
@johnbelamaric : are you ok with this last version (if yes, please remove the "requested changes" flag)
@miekg : I updated after latests comments (and remove the list of security team in OWNERS). Let me know if ok. 
Thanks.

**Review by johnbelamaric — APPROVED** [review]

File: OWNERS
Comment (miekg) [inline OWNERS]: this file shouldn't be touched by this PR anymore

**Review by miekg — COMMENTED** [review]
for the rest lgtm

**Comment by fturib** [comment]
I updated. Just reverting the OWNER file that is no more involved in this PR.

**Pull request merged**
